curvefs/src/client/volume/default_volume_storage.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Date: Monday Mar 14 17:40:21 CST 2022
+ * Author: wuhanqing
+ */
+
+#include "curvefs/src/client/volume/default_volume_storage.h"
+
+#include <bvar/bvar.h>
+#include <glog/logging.h>
+
+#include <memory>
+#include <sstream>
+#include <type_traits>
+#include <vector>
+
+#include "absl/meta/type_traits.h"
+#include "curvefs/src/client/inode_cache_manager.h"
+#include "curvefs/src/client/inode_wrapper.h"
+#include "curvefs/src/client/volume/extent_cache.h"
+#include "curvefs/src/client/volume/utils.h"
+
+namespace curvefs {
+namespace client {
+
+namespace {
+
+template <typename IOPart,
+          typename = absl::void_t<decltype(std::declval<IOPart>().offset),
+                                  decltype(std::declval<IOPart>().length)>>
+std::ostream& operator<<(std::ostream& os, const std::vector<IOPart>& iov) {
+    if (iov.empty()) {
+        os << "empty";
+        return os;
+    }
+
+    std::ostringstream oss;
+    oss << "{";
+    for (const auto& io : iov) {
+        oss << io.offset << "~" << io.length << ",";
+    }
+
+    oss << "}";
+    os << oss.str();
+
+    return os;
+}
+
+}  // namespace
+
+ssize_t DefaultVolumeStorage::Write(uint64_t ino,
+                                    off_t offset,
+                                    size_t len,
+                                    const char* data) {
+    std::shared_ptr<InodeWrapper> inodeWrapper;
+    LatencyUpdater updater(&metric_.writeLatency);
+    auto ret = inodeCacheManager_->GetInode(ino, inodeWrapper);
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Get inode error, ino: " << ino << ", ret: " << ret;
+        return static_cast<ssize_t>(ret);
+    }
+
+    auto extentCache = inodeWrapper->GetMutableExtentCache();
+
+    std::vector<WritePart> writes;
+    if (!PrepareWriteRequest(offset, len, data, extentCache, spaceManager_,
+                             &writes)) {
+        LOG(ERROR) << "Prepare write requests error, ino: " << ino
+                   << ", offset: " << offset << ", len: " << len;
+        return static_cast<ssize_t>(CURVEFS_ERROR::IO_ERROR);
+    }
+
+    VLOG(9) << "write ino: " << ino << ", offset: " << offset
+            << ", len: " << len << ", block write requests: " << writes;
+
+    ssize_t nr = blockDeviceClient_->Writev(writes);
+    if (nr < 0 /*|| nr != len*/) {
```

Comment:
Why comment out the latter part： nr != len

Reply:
for unaligned requests, the return value will be bigger than the origin read/write length.

Reply:
remark here
<img width="605" alt="image" src="https://user-images.githubusercontent.com/15689619/161885783-acf67321-1678-4ccf-a618-51711799231f.png">